File: EnviroStatus_VB/Zrodlo_GIOS.vb
Public Class Source_GIOS
    Inherits Source_Base

    Protected Overrides Property SRC_SETTING_NAME As String = "sourceGIOS"
    Protected Overrides Property SRC_SETTING_HEADER As String = "GIOŚ"
    Protected Overrides Property SRC_RESTURI_BASE As String = "http://api.gios.gov.pl/pjp-api/rest/"
    Public Overrides ReadOnly Property SRC_POMIAR_SOURCE As String = "gios"
    Protected Overrides Property SRC_HAS_TEMPLATES As Boolean = True
    Public Overrides ReadOnly Property SRC_IN_TIMER As Boolean = True

    ' API: http://powietrze.gios.gov.pl/pjp/content/api
    ' LIMIT: 2x na godzinę

    Private Sub AddPomiar(oNew As JedenPomiar)
        For Each oItem As JedenPomiar In moListaPomiarow
            If oItem.sPomiar = oNew.sPomiar Then
                ' porownanie dat

                ' porownanie odleglosci
                If oItem.dOdl > oNew.dOdl Then
                    ' moListaPomiarow.Remove(oItem)
                    oItem.bDel = True
                    ' oNew zostanie dodany po zakonczeniu petli
                Else
                    Exit Sub    ' mamy nowszy pomiar, czyli oNew nas nie interesuje
                End If
            End If
        Next
        moListaPomiarow.Add(oNew)

    End Sub

    Private Function NormalizePomiarName(sPomiar As String) As String

        Select Case sPomiar
            Case "C6H6"
                Return "C₆H₆"
            Case "SO2"
                Return "SO₂"
            Case "NO2"
                Return "NO₂"
            Case "O3"
                Return "O₃"
            Case "PM10"
                Return "PM₁₀"
            Case "PM1"
                Return "PM₁"
            Case "PM2.5"
                Return "PM₂₅"
        End Select
        Return sPomiar
    End Function

    Private Function Unit4Pomiar(sPomiar As String) As String
        Return " μg/m³"
    End Function


    Private Async Function GetPomiary(oTemplate As JedenPomiar, bInTimer As Boolean) As Task
        ' do moListaPomiarow dodaj wszystkie pomiary robione przez oTemplate.sId

        Dim sCmd As String
        sCmd = "station/sensors/" & oTemplate.sId
        Dim sPage As String = Await GetREST(sCmd)

        Dim bError As Boolean = False
        Dim oJson As Windows.Data.Json.JsonArray = Nothing
        Try
            oJson = Windows.Data.Json.JsonArray.Parse(sPage)
        Catch ex As Exception
            bError = True
        End Try
        If bError Then
            If Not bInTimer Then Await DialogBox("ERROR: JSON parsing error - gios/GetPomiary")
            Exit Function
        End If

        Try

            For Each oJsonMeasurement As Windows.Data.Json.JsonValue In oJson

                Dim oNew As JedenPomiar = New JedenPomiar With {
                .sSource = oTemplate.sSource,
                .sId = oTemplate.sId,
                .dLon = oTemplate.dLon,
                .dLat = oTemplate.dLat,
                .dWysok = oTemplate.dWysok,
                .dOdl = oTemplate.dOdl,
                .sOdl = Odleglosc2String(oTemplate.dOdl),
                .sSensorDescr = oTemplate.sSensorDescr,
                .sAdres = oTemplate.sAdres
                }
                ' .sTimeStamp = oTemplate.sTimeStamp

                oNew.sAddit = oJsonMeasurement.GetObject().GetNamedNumber("id")

                Dim oJsonVal As Windows.Data.Json.JsonValue
                oJsonVal = oJsonMeasurement.GetObject().GetNamedValue("param")

                oNew.sPomiar = oJsonVal.GetObject().GetNamedString("paramCode")
                AddPomiar(oNew)
            Next
        Catch ex As Exception
            ' w razie bledu - nic
        End Try


        '  [{
        '    "id": 92,
        '    "stationId": 14,
        '    "param": {
        '        "paramName": "pył zawieszony PM10",
        '        "paramFormula": "PM10",
        '        "paramCode": "PM10",
        '        "idParam": 3
        '    }
        '}, ...

    End Function

    Private Async Function GetWartosci() As Task
        Try

            For Each oItem As JedenPomiar In moListaPomiarow
                If oItem.bDel Then Continue For

                Dim sCmd As String
                sCmd = "data/getData/" & oItem.sAddit
                Dim sPage As String = Await GetREST(sCmd)

                Dim bError As Boolean = False
                Dim oJson As Windows.Data.Json.JsonValue = Nothing
                Try
                    oJson = Windows.Data.Json.JsonValue.Parse(sPage)
                Catch ex As Exception
                    bError = True
                End Try
                If bError Then
                    Await DialogBox("ERROR: JSON parsing error - GIOS/GetWartosci")
                    Return
                End If

                Dim oJsonArr As Windows.Data.Json.JsonArray
                oJsonArr = oJson.GetObject().GetNamedArray("values")

                For Each oJsonMeasurement As Windows.Data.Json.JsonValue In oJsonArr
                    oItem.sTimeStamp = oJsonMeasurement.GetObject().GetNamedString("date")
                    oItem.sCurrValue = ""
                    Try
                        Dim oVal As Windows.Data.Json.JsonValue
                        oVal = oJsonMeasurement.GetObject().GetNamedValue("value")
                        If oVal.ValueType <> Windows.Data.Json.JsonValueType.Null Then
                            oItem.sCurrValue = oJsonMeasurement.GetObject().GetNamedNumber("value")
                        End If
                    Catch ex As Exception
                        ' bo moze byc NULL :)
                    End Try
                    If oItem.sCurrValue <> "" Then
                        oItem.dCurrValue = oItem.sCurrValue
                        oItem.sCurrValue = oItem.dCurrValue
                        oItem.sUnit = Unit4Pomiar(oItem.sPomiar)
                        If oItem.sCurrValue.Length > 5 Then oItem.sCurrValue = oItem.sCurrValue.Substring(0, 5)
                        oItem.sCurrValue = oItem.sCurrValue & oItem.sUnit
                        Exit For
                    End If
                Next
                If oItem.sCurrValue = "" Then
                    oItem.bDel = True
                    oItem.dCurrValue = 0
                Else
                    oItem.sPomiar = NormalizePomiarName(oItem.sPomiar)
                End If
            Next
        Catch ex As Exception
            'w razie bledu - nic. 
        End Try
    End Function

    Public Overrides Async Function GetDataFromFavSensor(sId As String, sAddit As String, bInTimer As Boolean) As Task(Of Collection(Of JedenPomiar))
        ' ale w efekcie jest kilka GIOSów jednego parametru
        moListaPomiarow = New Collection(Of JedenPomiar)
        If Not GetSettingsBool("sourceGIOS", True) Then Return moListaPomiarow

        Dim oTemplate As JedenPomiar = New JedenPomiar

        ' wczytaj dane template dla danego favname
        Dim oFile As Windows.Storage.StorageFile =
        Await App.GetDataFile(False, "gios_" & sId & ".xml", False)
        If oFile IsNot Nothing Then
            Dim oSer As Xml.Serialization.XmlSerializer =
                New Xml.Serialization.XmlSerializer(GetType(JedenPomiar))
            Dim oStream As Stream = Await oFile.OpenStreamForReadAsync
            oTemplate = TryCast(oSer.Deserialize(oStream), JedenPomiar)
            oStream.Dispose()   ' == fclose
        Else
            oTemplate = New JedenPomiar
        End If

        oTemplate.sSource = SRC_POMIAR_SOURCE  ' to tak na wszelki wypadek
        oTemplate.sId = sId
        'oTemplate.dLon = oJsonObj.GetNamedstring("gegrLon")
        'oTemplate.dLat = oJsonObj.GetNamedstring("gegrLat")
        'oTemplate.dWysok = 0    ' brak danych
        'oTemplate.dOdl = App.GPSdistanceDwa(oPos.X, oPos.Y,
        '                                    oTemplate.dLat, oTemplate.dLon)
        'oTemplate.sOdl = oTemplate.dOdl & " m"
        'oTemplate.sSensorDescr = oJsonObj.GetNamedString("stationName")
        'oTemplate.sAdres = oJsonObj.GetNamedString("addressStreet")


        Await GetPomiary(oTemplate, bInTimer)

        ' teraz odczytaj wartosci!
        Await GetWartosci()

        Return moListaPomiarow

    End Function

    Public Overrides Async Function GetNearest(oPos As Point) As Task(Of Collection(Of JedenPomiar))
        Dim dMaxOdl As Double = 10

        moListaPomiarow = New Collection(Of JedenPomiar)
        If Not GetSettingsBool("sourceGIOS", True) Then Return moListaPomiarow

        Dim sCmd As String
        sCmd = "station/findAll"
        Dim sPage As String = Await GetREST(sCmd)


        Dim bError As Boolean = False
        Dim oJson As Windows.Data.Json.JsonArray = Nothing
        Try
            oJson = Windows.Data.Json.JsonArray.Parse(sPage)
        Catch ex As Exception
            bError = True
        End Try
        If bError Then
            Await DialogBox("ERROR: JSON parsing error - GIOS\sPage")
            Return moListaPomiarow
        End If

        Try

            If oJson.Count = 0 Then Return moListaPomiarow     ' brak bliskich?


            ' przeiteruj, tworząc JedenPomiar dla kazdego pomiaru i kazdej lokalizacji

            For Each oJsonSensor As Windows.Data.Json.JsonValue In oJson
                Dim oJsonObj As Windows.Data.Json.JsonObject
                oJsonObj = oJsonSensor.GetObject()

                Dim oTemplate As JedenPomiar = New JedenPomiar
                oTemplate.sSource = SRC_POMIAR_SOURCE
                oTemplate.sId = oJsonObj.GetNamedNumber("id")

                oTemplate.dLon = oJsonObj.GetNamedString("gegrLon")
                oTemplate.dLat = oJsonObj.GetNamedString("gegrLat")

                oTemplate.dWysok = 0    ' brak danych

                oTemplate.dOdl = GPSdistanceDwa(oPos.X, oPos.Y,
                                                oTemplate.dLat, oTemplate.dLon)

                If oTemplate.dOdl / 1000 < dMaxOdl Then
                    ' teraz cos, co chce dodac

                    oTemplate.sOdl = Odleglosc2String(oTemplate.dOdl)

                    oTemplate.sSensorDescr = oJsonObj.GetNamedString("stationName")

                    oTemplate.sAdres = oJsonObj.GetNamedString("addressStreet")
                    If oTemplate.sAdres = "" Then
                        Try
                            Dim oJsonAdres As Windows.Data.Json.JsonObject
                            oJsonAdres = oJsonObj.GetNamedValue("city").GetObject()
                            oTemplate.sAdres = oJsonAdres.GetNamedString("name")
                            Dim oJsonComm As Windows.Data.Json.JsonObject
                            oJsonComm = oJsonAdres.GetNamedValue("commune").GetObject()
                            oTemplate.sAdres = oTemplate.sAdres & vbCrLf & "("
                            If oJsonComm.GetNamedString("communeName") <> "" Then
                                oTemplate.sAdres = oTemplate.sAdres & "gmina " & oJsonComm.GetNamedString("communeName") & vbCrLf
                            End If
                            If oJsonComm.GetNamedString("districtName") <> "" Then
                                oTemplate.sAdres = oTemplate.sAdres & "powiat " & oJsonComm.GetNamedString("districtName") & vbCrLf
                            End If
                            If oJsonComm.GetNamedString("provinceName") <> "" Then
                                oTemplate.sAdres = oTemplate.sAdres & oJsonComm.GetNamedString("provinceName") & vbCrLf
                            End If
                        Catch ex As Exception
                        End Try


                    End If

                    Await GetPomiary(oTemplate, False)
                End If
            Next

            ' teraz odczytaj wartosci!
            Await GetWartosci()
        Catch ex As Exception

        End Try

        Return moListaPomiarow
        ' {""id"":14,""stationName"":""Działoszyn"",""gegrLat"":""50.972167"",""gegrLon"":""14.941319"",""city"":{""id"":192,""name"":""Działoszyn"",""commune"":{""communeName"":""Bogatynia"",""districtName"":""zgorzelecki"",""provinceName"":""DOLNOŚLĄSKIE""}},""addressStreet"":null},
        ' {""id"":400,""stationName"":""Kraków, Aleja Krasińskiego"",""gegrLat"":""50.057678"",""gegrLon"":""19.926189"",""city"":{""id"":415,""name"":""Kraków"",""commune"":{""communeName"":""Kraków"",""districtName"":""Kraków"",""provinceName"":""MAŁOPOLSKIE""}},""addressStreet"":""al. Krasińskiego""}
    End Function



End Class

File: EnviroStatus/EnviroStatus.Shared/Zrodlo_GIOS.cs
using Windows.Foundation;
//using System.Threading.Tasks;
using System.IO;
//using Microsoft.VisualBasic;
using System.Collections.ObjectModel;
using System;
//using Microsoft.VisualBasic.CompilerServices;

namespace EnviroStatus
{
    public class Source_GIOS : Source_Base
    {
        protected override string SRC_SETTING_NAME { get;  } = "sourceGIOS";
        protected override string SRC_SETTING_HEADER { get;  } = "GIOŚ";
        protected override string SRC_RESTURI_BASE { get;  } = "http://api.gios.gov.pl/pjp-api/rest/";
        public override string SRC_POMIAR_SOURCE { get;  } = "gios";
        protected override bool SRC_HAS_TEMPLATES { get;  } = true;
        public override bool SRC_IN_TIMER { get;  } = true;
        protected override string SRC_URI_ABOUT_EN { get; } = "http://powietrze.gios.gov.pl/pjp/current";
        protected override string SRC_URI_ABOUT_PL { get; } = "http://powietrze.gios.gov.pl/pjp/current";

        // API: http://powietrze.gios.gov.pl/pjp/content/api
        // LIMIT: 2x na godzinę

        private void AddPomiar(JedenPomiar oNew)
        {
            foreach (EnviroStatus.JedenPomiar oItem in moListaPomiarow)
            {
                if ((oItem.sPomiar ?? "") == (oNew.sPomiar ?? ""))
                {
                    // porownanie dat

                    // porownanie odleglosci
                    if (oItem.dOdl > oNew.dOdl)
                        // moListaPomiarow.Remove(oItem)
                        oItem.bDel = true;
                    else
                        return;// mamy nowszy pomiar, czyli oNew nas nie interesuje
                }
            }
            moListaPomiarow.Add(oNew);
        }

        private string NormalizePomiarName(string sPomiar)
        {
            switch (sPomiar)
            {
                case "C6H6":
                        return "C₆H₆";
                case "SO2":
                        return "SO₂";
                case "NO2":
                        return "NO₂";
                case "O3":
                        return "O₃";
                case "PM10":
                        return "PM₁₀";
                case "PM1":
                        return "PM₁";
                case "PM2.5":
                        return "PM₂₅";
            }
            return sPomiar;
        }

        private string Unit4Pomiar(string sPomiar)
        {
            return " μg/m³";
        }


        private async System.Threading.Tasks.Task GetPomiary(JedenPomiar oTemplate, bool bInTimer)
        {
            // do moListaPomiarow dodaj wszystkie pomiary robione przez oTemplate.sId

            string sCmd;
            sCmd = "station/sensors/" + oTemplate.sId;
            string sPage = await GetREST(sCmd);
            if (sPage.Length < 10)
                return;

            bool bError = false;
            string sErr = "";
            Newtonsoft.Json.Linq.JArray oJson = null;
            try
            {
                oJson = Newtonsoft.Json.Linq.JArray.Parse(sPage);
            }
            catch (Exception ex)
            {
                sErr = ex.Message;
                bError = true;
            }
            if (bError)
            {
                if (!bInTimer)
                    await p.k.DialogBoxAsync("ERROR: JSON parsing error - getting sensor data (" + SRC_POMIAR_SOURCE + ")\n " + sErr);
                return;
            }

            try
            {
                foreach (Newtonsoft.Json.Linq.JToken oJsonMeasurement in oJson)
                {
                    var oNew = new JedenPomiar()
                    {
                        sSource = oTemplate.sSource,
                        sId = oTemplate.sId,
                        dLon = oTemplate.dLon,
                        dLat = oTemplate.dLat,
                        dWysok = oTemplate.dWysok,
                        dOdl = oTemplate.dOdl,
                        sOdl = Odleglosc2String(oTemplate.dOdl),
                        sSensorDescr = oTemplate.sSensorDescr,
                        sAdres = oTemplate.sAdres
                    };
                    // .sTimeStamp = oTemplate.sTimeStamp

                    oNew.sAddit = oJsonMeasurement.GetObject().GetNamedNumber("id").ToString();

                    Newtonsoft.Json.Linq.JToken oJsonVal;
                    oJsonVal = oJsonMeasurement.GetObject().GetNamedToken("param");

                    oNew.sPomiar = oJsonVal.GetObject().GetNamedString("paramCode");
                    AddPomiar(oNew);
                }
            }
            catch 
            {
            }

        }

        private async System.Threading.Tasks.Task GetWartosci()
        {
            try
            {
                foreach (EnviroStatus.JedenPomiar oItem in moListaPomiarow)
                {
                    if (oItem.bDel)
                        continue;

                    string sCmd;
                    sCmd = "data/getData/" + oItem.sAddit;
                    string sPage = await GetREST(sCmd);
                    if (sPage.Length < 10)
                        return;

                    bool bError = false;
                    string sErr = "";
                    Newtonsoft.Json.Linq.JObject oJson = null;
                    try
                    {
                        oJson = Newtonsoft.Json.Linq.JObject.Parse(sPage);
                    }
                    catch (Exception ex)
                    {
                        sErr = ex.Message;
                        bError = true;
                    }
                    if (bError)
                    {
                        await p.k.DialogBoxAsync("ERROR: JSON parsing error - getting values (" + SRC_POMIAR_SOURCE + ")\n " + sErr);
                        return;
                    }

                    Newtonsoft.Json.Linq.JArray oJsonArr;
                    oJsonArr = oJson.GetObject().GetNamedArray("values");

                    foreach (Newtonsoft.Json.Linq.JToken oJsonMeasurement in oJsonArr)
                    {
                        oItem.sTimeStamp = oJsonMeasurement.GetObject().GetNamedString("date");
                        oItem.sCurrValue = "";
                        try
                        {
                            Newtonsoft.Json.Linq.JToken oVal;
                            oVal = oJsonMeasurement.GetObject().GetNamedToken("value");
                            if (oVal.Type != Newtonsoft.Json.Linq.JTokenType.Null )
                                oItem.sCurrValue = oJsonMeasurement.GetObject().GetNamedNumber("value").ToString();
                        }
                        catch 
                        {
                        }
                        if (!string.IsNullOrEmpty(oItem.sCurrValue))
                        {
                            oItem.dCurrValue = oItem.sCurrValue.ParseDouble(0);
                            // oItem.sCurrValue = Conversions.ToString(oItem.dCurrValue);
                            oItem.sUnit = Unit4Pomiar(oItem.sPomiar);
                            if (oItem.sCurrValue.Length > 5)
                                oItem.sCurrValue = oItem.sCurrValue.Substring(0, 5);
                            oItem.sCurrValue = oItem.sCurrValue + oItem.sUnit;
                            break;
                        }
                    }
                    if (string.IsNullOrEmpty(oItem.sCurrValue))
                    {
                        oItem.bDel = true;
                        oItem.dCurrValue = 0;
                    }
                    else
                        oItem.sPomiar = NormalizePomiarName(oItem.sPomiar);
                }
            }
            catch 
            {
            }


        }

        public override async System.Threading.Tasks.Task<Collection<EnviroStatus.JedenPomiar>> GetDataFromFavSensor(string sId, string sAddit, bool bInTimer)
        {
            // ale w efekcie jest kilka GIOSów jednego parametru
            moListaPomiarow = new Collection<JedenPomiar>();
            if (!p.k.GetSettingsBool("sourceGIOS", SRC_DEFAULT_ENABLE))
                return moListaPomiarow;

            var oTemplate = new JedenPomiar();

            // wczytaj dane template dla danego favname
            var oFile = await EnviroStatus.App.GetDataFile(false, "gios_" + sId + ".xml", false);
            if (oFile != null)
            {
                var oSer = new System.Xml.Serialization.XmlSerializer(typeof(JedenPomiar));
                var oStream = await oFile.OpenStreamForReadAsync();
                oTemplate = oSer.Deserialize(oStream) as JedenPomiar;
                oStream.Dispose();   // == fclose
            }
            else
                oTemplate = new JedenPomiar();

            oTemplate.sSource = SRC_POMIAR_SOURCE;  // to tak na wszelki wypadek
            oTemplate.sId = sId;
            // oTemplate.dLon = oJsonObj.GetNamedstring("gegrLon")
            // oTemplate.dLat = oJsonObj.GetNamedstring("gegrLat")
            // oTemplate.dWysok = 0    ' brak danych
            // oTemplate.dOdl = App.GPSdistanceDwa(oPos.X, oPos.Y,
            // oTemplate.dLat, oTemplate.dLon)
            // oTemplate.sOdl = oTemplate.dOdl & " m"
            // oTemplate.sSensorDescr = oJsonObj.GetNamedString("stationName")
            // oTemplate.sAdres = oJsonObj.GetNamedString("addressStreet")


            await GetPomiary(oTemplate, bInTimer);

            // teraz odczytaj wartosci!
            await GetWartosci();

            return moListaPomiarow;
        }

        public override async System.Threading.Tasks.Task<Collection<EnviroStatus.JedenPomiar>> GetNearest(Windows.Devices.Geolocation.BasicGeoposition oPos)
        {
            moListaPomiarow = new Collection<JedenPomiar>();
            if (!p.k.GetSettingsBool("sourceGIOS", SRC_DEFAULT_ENABLE))
                return moListaPomiarow;

            if (!oPos.IsInsidePoland())
                return moListaPomiarow;

            double dMaxOdl = 10;

            string sCmd;
            sCmd = "station/findAll";
            string sPage = await GetREST(sCmd);
            if (sPage.Length < 10)
                return moListaPomiarow;

            bool bError = false;
            string sErr = "";
            Newtonsoft.Json.Linq.JArray oJson = null;
            try
            {
                oJson = Newtonsoft.Json.Linq.JArray.Parse(sPage);
            }
            catch (Exception ex)
            {
                sErr = ex.Message;
                bError = true;
            }
            if (bError)
            {
                await p.k.DialogBoxAsync("ERROR: JSON parsing error - getting nearest (" + SRC_POMIAR_SOURCE + ")\n " + sErr);
                return moListaPomiarow;
            }

            try
            {
                if (oJson.Count == 0)
                    return moListaPomiarow;     // brak bliskich?


                // przeiteruj, tworząc JedenPomiar dla kazdego pomiaru i kazdej lokalizacji

                foreach (Newtonsoft.Json.Linq.JToken oJsonSensor in oJson)
                {
                    Newtonsoft.Json.Linq.JToken oJsonObj;
                    oJsonObj = oJsonSensor.GetObject();

                    var oTemplate = new JedenPomiar();
                    oTemplate.sSource = SRC_POMIAR_SOURCE;
                    oTemplate.sId = oJsonObj.GetNamedNumber("id").ToString();

                    oTemplate.dLon = oJsonObj.GetNamedString("gegrLon").ParseDouble(0);
                    oTemplate.dLat = oJsonObj.GetNamedString("gegrLat").ParseDouble(0);

                    oTemplate.dWysok = 0;    // brak danych

                    oTemplate.dOdl = (int)oPos.DistanceTo(oTemplate.dLat, oTemplate.dLon);

                    if (oTemplate.dOdl / 1000 < dMaxOdl)
                    {
                        // teraz cos, co chce dodac

                        oTemplate.sOdl = Odleglosc2String(oTemplate.dOdl);

                        oTemplate.sSensorDescr = oJsonObj.GetNamedString("stationName");

                        oTemplate.sAdres = oJsonObj.GetNamedString("addressStreet");
                        if (string.IsNullOrEmpty(oTemplate.sAdres))
                        {
                            try
                            {
                                Newtonsoft.Json.Linq.JToken oJsonAdres;
                                oJsonAdres = oJsonObj.GetNamedToken("city");
                                oTemplate.sAdres = oJsonAdres.GetNamedString("name");
                                Newtonsoft.Json.Linq.JToken oJsonComm;
                                oJsonComm = oJsonAdres.GetNamedToken("commune").GetObject();
                                oTemplate.sAdres = oTemplate.sAdres + "\n(";
                                if (!string.IsNullOrEmpty(oJsonComm.GetNamedString("communeName")))
                                    oTemplate.sAdres = oTemplate.sAdres + "gmina " + oJsonComm.GetNamedString("communeName") + "\n";
                                if (!string.IsNullOrEmpty(oJsonComm.GetNamedString("districtName")))
                                    oTemplate.sAdres = oTemplate.sAdres + "powiat " + oJsonComm.GetNamedString("districtName") + "\n";
                                if (!string.IsNullOrEmpty(oJsonComm.GetNamedString("provinceName")))
                                    oTemplate.sAdres = oTemplate.sAdres + oJsonComm.GetNamedString("provinceName") + "\n";
                            }
                            catch 
                            {
                            }
                        }

                        await GetPomiary(oTemplate, false);
                    }
                }

                // teraz odczytaj wartosci!
                await GetWartosci();
            }
            catch 
            {
            }

            return moListaPomiarow;
        }
    }
}
